src/core/remote_io.c

```diff
@@ -0,0 +1,513 @@
+    if (backend_cnt != frontend_cnt) {
+        ERROR("There is no 1-to-1 mapping between a Remote I/O backend and Remote I/O frontend");
+    }
+
+    /** Update the Remote I/O device configuration */
+    for (vm_id = 0; vm_id < config.vmlist_size; vm_id++) {
+        struct vm_config* vm_config = &config.vmlist[vm_id];
+        for (i = 0; i < vm_config->platform.remote_io_dev_num; i++) {
+            struct remote_io_dev* dev = &vm_config->platform.remote_io_devs[i];
+            list_foreach (remote_io_device_list, struct remote_io_device, io_device) {
+                if (dev->id == io_device->id) {
+                    if (dev->is_backend) {
+                        io_device->config.backend_vm_id = vm_id;
+                        io_device->config.backend_interrupt = dev->interrupt;
+                        io_device->config.backend_cpu_id = (cpuid_t)-1;
+                    } else {
+                        io_device->config.frontend_vm_id = vm_id;
+                        io_device->config.frontend_interrupt = dev->interrupt;
+                        io_device->config.frontend_cpu_id = (cpuid_t)-1;
+                    }
+                }
+            }
+        }
+    }
+
+    /** Initialize the Remote I/O requests array */
+    for (size_t cpu_idx = 0; cpu_idx < REMOTE_IO_CPU_NUM; cpu_idx++) {
+        for (size_t vcpu_idx = 0; vcpu_idx < REMOTE_IO_VCPU_NUM; vcpu_idx++) {
+            remote_io_requests[cpu_idx][vcpu_idx].state = REMOTE_IO_STATE_FREE;
+        }
+    }
+}
+
+void remote_io_assign_cpus(struct vm* vm)
+{
+    list_foreach (remote_io_device_list, struct remote_io_device, io_device) {
+        if (vm->id == io_device->config.backend_vm_id) {
+            io_device->config.backend_cpu_id = MIN(io_device->config.backend_cpu_id, cpu()->id);
+        } else if (vm->id == io_device->config.frontend_vm_id) {
+            io_device->config.frontend_cpu_id = MIN(io_device->config.frontend_cpu_id, cpu()->id);
+        }
+    }
+}
+
+/**
+ * @brief Performs the write or read operation by updating the request value
+ * @param id Remote I/O device ID
+ * @param addr MMIO register address
+ * @param value Value to be written or read
+ * @param cpu_id CPU ID of the frontend VM that issued the I/O request
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns true if the operation was successful, false otherwise
+ */
+static bool remote_io_w_r_operation(unsigned long id, unsigned long addr, unsigned long value,
+    unsigned long cpu_id, unsigned long vcpu_id)
+{
+    list_foreach (remote_io_device_list, struct remote_io_device, io_device) {
+        if (io_device->id == id) {
+            spin_lock(&remote_io_request_lock);
+            struct remote_io_request* node = &remote_io_requests[cpu_id][vcpu_id];
+            spin_unlock(&remote_io_request_lock);
+
+            if (node->addr != addr || node->state != REMOTE_IO_STATE_PROCESSING) {
+                break;
+            }
+
```

Comment:
The lock is only required to prevent the backend VM from extracting the I/O requests from the slot while the frontend VM is still inserting them. However, this race condition only occurs when the backend VM operates in polling mode. When driven by interrupts, we ensure that the request field is fully updated in the I/O request array slot before raising the interrupt.

Since the I/O request slot (or node) is managed via a finite state machine (FSM) and updated sequentially, no lock is necessary. The design ensures there is no concurrent access to the same variable, even when the frontend and backend are running on multiple vCPUs.